### PR TITLE
recover persistent storage from snapshot file

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
add recover logic of maintenance snapshot file
* what changes does this pull request make?
implement a tmp client to restore, will be merge in our own client in the future
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/client/restore.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#[inline]
+#[allow(clippy::indexing_slicing)] // safe operation
+pub async fn restore(
+    snapshot_path: impl AsRef<Path>,
+    data_dir: impl AsRef<Path>,
+) -> Result<(), ClientError> {
+    let mut snapshot_f = tokio::fs::File::open(snapshot_path).await?;
+    let tmp_path = format!("/tmp/snapshot-{}", uuid::Uuid::new_v4());
+    let mut rocks_snapshot = RocksSnapshot::new_for_receiving(tmp_path.clone())?;
+    let mut buf = vec![0; 8192];
```
Comment (Phoenix500526) [inline xline/src/client/restore.rs:24]: Although we can use different buffers with different sizes to store the data read from a snapshot, I still recommend using a unified size to reduce some confusion. Can we use the `MAINTENANCE_SNAPSHOT_CHUNK_SIZE` here? In doing so, we don't need to leave a magic number here or define a constant variable to replace it.

File: xline/src/client/restore.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    while let Ok(n) = snapshot_f.read(&mut buf).await {
+        if n == 0 {
+            break;
+        }
+        let _ignore = rocks_snapshot.write(&buf[..n]).await?;
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/client/restore.rs:29]: snapshot may be a large file, using `write_all` needs to read it completely into memory

[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/client/restore.rs:32]: Fixed

File: xline/src/client/restore.rs (hunk 1 above)
Comment (themanforfree) [reply xline/src/client/restore.rs]: Fixed

File: xline/src/client/restore.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Phoenix500526) [reply xline/src/client/restore.rs:32]: Doesn't `write_all` means write all data of the given buffer into a snapshot? That seems how we use it in the `CurpNode::install_snapshot`.

[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/client/restore.rs]: Sry, my mistake, fixed

**Review by Phoenix500526 — DISMISSED** [review]

File: engine/src/rocksdb_engine.rs [hunk 3]
```diff
@@ -279,6 +279,11 @@ impl RocksSnapshot {
     }
 
     /// Write snapshot data
+    /// # Errors
+    /// Return `io::Error` when write failed with io error.
+    /// # Panics
```
Comment (rogercloud) [inline engine/src/rocksdb_engine.rs:284]: Shall we not panic when got a invalid snapshot file size? I think report an error is better.

Comment (themanforfree) [reply engine/src/rocksdb_engine.rs]: Fixed

**Review by rogercloud — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]

**Pull request merged**